[llvm-readobj][AArch64][ELF][PAC] Support ELF AUTH constants

kovdan01 opened the pull request:
This patch adds llvm-readobj support for:

- Dynamic R_AARCH64_AUTH_* relocations (including RELR compressed AUTH relocations) as described here: https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#auth-variant-dynamic-relocations

- .note.AARCH64-PAUTH-ABI-tag section as defined here https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#elf-marking

jh7370 commented on llvm/include/llvm/BinaryFormat/ELF.h:
```diff
@@ -1643,6 +1646,11 @@ enum {
   NT_ANDROID_TYPE_MEMTAG = 4,
 };
 
+// ARM note types
```
```suggestion
// ARM note types.
```
Nit

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/broken-dynamic-reloc.test:
```diff
@@ -138,19 +138,19 @@ ProgramHeaders:
 
-## INVALID-DT-RELRENT:         invalid DT_RELRSZ value (0x18) or DT_RELRENT value (0xff)
-## INVALID-DT-ANDROID-RELRENT: invalid DT_ANDROID_RELRSZ value (0x18) or DT_ANDROID_RELRENT value (0xff)
+## INVALID-DT-RELRENT:              invalid DT_RELRSZ value (0x18) or DT_RELRENT value (0xff)
+## INVALID-DT-ANDROID-RELRENT:      invalid DT_ANDROID_RELRSZ value (0x18) or DT_ANDROID_RELRENT value (0xff)
```
Why have these two  lines changed?

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/broken-dynamic-reloc.test:
```diff
@@ -138,19 +138,19 @@ ProgramHeaders:
 
-## INVALID-DT-RELRSZ:         warning: '[[FILE]]': invalid DT_RELRSZ value (0xff) or DT_RELRENT value (0x18)
-## INVALID-DT-ANDROID-RELRSZ: warning: '[[FILE]]': invalid DT_ANDROID_RELRSZ value (0xff) or DT_ANDROID_RELRENT value (0x18)
+## INVALID-DT-RELRSZ:              warning: '[[FILE]]': invalid DT_RELRSZ value (0xff) or DT_RELRENT value (0x18)
+## INVALID-DT-ANDROID-RELRSZ:      warning: '[[FILE]]': invalid DT_ANDROID_RELRSZ value (0xff) or DT_ANDROID_RELRENT value (0x18)
```
Why have these twolines  changed?

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/broken-dynamic-reloc.test:
```diff
@@ -138,19 +138,19 @@ ProgramHeaders:
-# RUN: yaml2obj --docnum=2 -DRELTYPE=RELR -DTAG1=DT_ANDROID_RELRSZ -DTAG2=DT_ANDROID_RELRENT -DTAG2VAL=0xFF %s -o %t9
-# RUN: llvm-readobj --dyn-relocations %t9 2>&1 | FileCheck %s -DFILE=%t9 --check-prefix=INVALID-DT-ANDROID-RELRENT
-# RUN: llvm-readelf --dyn-relocations %t9 2>&1 | FileCheck %s -DFILE=%t9 --check-prefix=INVALID-DT-ANDROID-RELRENT
+# RUN: yaml2obj --docnum=2 -DRELTYPE=RELR -DTAG1=DT_RELRSZ -DTAG2=DT_RELRENT -DTAG2VAL=0xFF %s -o %t9
+# RUN: llvm-readobj --dyn-relocations %t9 2>&1 | FileCheck %s -DFILE=%t9 --check-prefix=INVALID-DT-RELRENT
+# RUN: llvm-readelf --dyn-relocations %t9 2>&1 | FileCheck %s -DFILE=%t9 --check-prefix=INVALID-DT-RELRENT
+# RUN: yaml2obj --docnum=2 -DRELTYPE=RELR -DTAG1=DT_ANDROID_RELRSZ -DTAG2=DT_ANDROID_RELRENT -DTAG2VAL=0xFF %s -o %t10
+# RUN: llvm-readobj --dyn-relocations %t10 2>&1 | FileCheck %s -DFILE=%t10 --check-prefix=INVALID-DT-ANDROID-RELRENT
+# RUN: llvm-readelf --dyn-relocations %t10 2>&1 | FileCheck %s -DFILE=%t10 --check-prefix=INVALID-DT-ANDROID-RELRENT
```
Please revert these chanhges: this file uses sequential numbering and you've added your new cases after them, so they shouldn't need updating.

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/broken-dynamic-reloc.test:
```diff
@@ -459,3 +459,43 @@ ProgramHeaders:
+# RUN: llvm-readobj --dyn-relocations %t8 2>&1 | FileCheck %s -DFILE=%t8 --check-prefix=INVALID-DT-AARCH64-AUTH-RELRSZ
+# RUN: llvm-readelf --dyn-relocations %t8 2>&1 | FileCheck %s -DFILE=%t8 --check-prefix=INVALID-DT-AARCH64-AUTH-RELRSZ
+
+## INVALID-DT-AARCH64-AUTH-RELRSZ: warning: '[[FILE]]': invalid DT_AARCH64_AUTH_RELRSZ value (0xff) or DT_AARCH64_AUTH_RELRENT value (0x18)
```
Comments use `##`. Lit and FileCheck directives should use single `#`. Also applies below.

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -2086,21 +2086,34 @@ template <typename ELFT> void ELFDumper<ELFT>::parseDynamicTable() {
-      DynRelrRegion.SizePrintName = Dyn.d_tag == ELF::DT_RELRSZ
-                                        ? "DT_RELRSZ value"
-                                        : "DT_ANDROID_RELRSZ value";
+      if (Dyn.d_tag == ELF::DT_RELRSZ)
```
Probably this should be another `switch`.

MaskRay replied:
For this case, I think it is better to do

```
if (emachine == EM_AARCH64 && xxx == ELF::DT_AARCH64_AUTH_RELRSZ) {
  DynRelrRegion.Size = Dyn.getVal();
  DynRelrRegion.SizePrintName = "DT_AARCH64_AUTH_RELRSZ value";
  continue;
}
switch (Dyn.d_tag) {
  ...
}
```

If emachine is not EM_AARCH64, `DT_AARCH64_AUTH_RELRSZ` is not defined.

jh7370 replied:
Agreed with @MaskRay.

kovdan01 replied:
Moved checks for `DT_AARCH64_AUTH_RELRSZ` and `DT_AARCH64_AUTH_RELRENT` to an `if` before the main `switch` as you suggested. These two values are distinguished inside the `if` via inner `switch` with `continue` at the end of each `case`.

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -2086,21 +2086,34 @@ template <typename ELFT> void ELFDumper<ELFT>::parseDynamicTable() {
-      DynRelrRegion.EntSizePrintName = Dyn.d_tag == ELF::DT_RELRENT
-                                           ? "DT_RELRENT value"
-                                           : "DT_ANDROID_RELRENT value";
+      if (Dyn.d_tag == ELF::DT_RELRENT)
```
Ditto.

kovdan01 replied:
See the comment above

jh7370 commented on llvm/test/tools/llvm-objdump/ELF/dynamic-section-machine-specific.test:
```diff
@@ -268,6 +268,9 @@ ProgramHeaders:
 # AARCH64:      Dynamic Section:
 # AARCH64-NEXT:  AARCH64_BTI_PLT      0x0000000000000001
 # AARCH64-NEXT:  AARCH64_PAC_PLT      0x0000000000000002
+# AARCH64-NEXT:  AARCH64_AUTH_RELR    0x0000000000000003
```
I wasn't expecting to see an llvm-objdump change in a patch about llvm-readobj?

jh7370 commented on llvm/include/llvm/BinaryFormat/ELF.h:
```diff
@@ -1643,6 +1646,11 @@ enum {
 
+// ARM note types
+enum {
+  NT_ARM_TYPE_PAUTH_ABI_TAG = 1,
```
I couldn't see this value defined in the linked spec?

jh7370 replied:
I think I have a few (small) concerns with this:
1) It's potentially possible (admittedly unlikely) that the name could be specified for something different in the future, so you'd have to modify existing sites to use a different name.
2) Giving it a name means people will search for the string in the docs and not find it, potentially causing confusion in the future.

I don't know who is in charge of those docs, but is it possible to update them to specifically name the NT_* value? That would resolve my concern.

kovdan01 replied:
@smithp35 Could we provide this constant name `NT_ARM_TYPE_PAUTH_ABI_TAG` (or some other name) and its value `1` in the section 14.1 of the spec https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#appendix-alternative-elf-marking-using-gnu-program-properties? I could submit a PR to https://github.com/ARM-software/abi-aa if you wish.

kovdan01 replied:
@smithp35 A kind reminder regarding my question above.

smithp35 replied:
Just to make sure I've got this right:
The link above is for GNU program properties:
https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#19appendix-alternative-elf-marking-using-gnu-program-properties

It defines an additional program property on top of the existing .note.GNU.property defined in the Linux ABI extensions https://raw.githubusercontent.com/wiki/hjl-tools/linux-abi/linux-abi-draft.pdf that already has its own `NT_GNU_PROPERTY_TYPE_0` 

There is a part here which defines .note.AARCH64-PAUTH-ABI-tag which I think is where an additional `NT_ARM_TYPE_PAUTH_ABI_TAG` would make sense.
https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#141default-marking-schema

Have I got that right?

In any case please can you raise an issue on the AArch64 ABI for that https://github.com/ARM-software/abi-aa so it can be addressed there. I'll also need to update the document to make it compatible with some forthcoming changes in the ABI to do with ELF marking. It is unfortunate that this will land too late for your timescale.

Essentially these pull-requests:
* https://github.com/ARM-software/abi-aa/pull/230 build attributes for relocatable objects
* https://github.com/ARM-software/abi-aa/pull/229 rationale for GNU properties in executable shared libraries

None have been merged yet, but pending more review comments there is a consensus within Arm that we'll be using build attributes (like 32-bit Arm) for ELF metadata in relocatable objects and for SysvAbi platforms such as Linux we'll be using .note.GNU.properties as our reference ELF marking scheme for executables and shared-libraries.

I expect that when the PAuthABI specification comes out of Alpha it will need to be updated, most likely to make the current Alternative ELF marking will be the reference marking, and the .note.AARCH64-PAUTH-ABI-tag will become the alternative. I've already defined build attibutes for PAuthABI that match the existing implementation.

kovdan01 replied:
> There is a part here which defines .note.AARCH64-PAUTH-ABI-tag which I think is where an additional `NT_ARM_TYPE_PAUTH_ABI_TAG` would make sense. https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#141default-marking-schema
> 
> Have I got that right?

Yes. It looks like I placed link a wrong link but a correct section number above :) I meant 14.1 https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#141default-marking-schema.

Submitted https://github.com/ARM-software/abi-aa/issues/234. Thanks!

jh7370 commented on llvm/include/llvm/BinaryFormat/ELFRelocs/AArch64.def:
```diff
@@ -121,6 +121,7 @@ ELF_RELOC(R_AARCH64_TLSLE_LDST128_TPREL_LO12,        0x23a)
 ELF_RELOC(R_AARCH64_TLSLE_LDST128_TPREL_LO12_NC,     0x23b)
 ELF_RELOC(R_AARCH64_TLSLD_LDST128_DTPREL_LO12,       0x23c)
 ELF_RELOC(R_AARCH64_TLSLD_LDST128_DTPREL_LO12_NC,    0x23d)
+ELF_RELOC(R_AARCH64_AUTH_ABS64,                      0x244)
```
I may be missing something but this value and the below one don't seem to line up with the 0xe100 and 0xe200 values in the spec?

kovdan01 committed:
[llvm-readobj][AArch64][ELF][PAC] Support ELF AUTH constants

This patch adds llvm-readobj support for:

- Dynamic R_AARCH64_AUTH_* relocations (including RELR compressed AUTH
  relocations) as described here:
  https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#auth-variant-dynamic-relocations

- .note.AARCH64-PAUTH-ABI-tag section as defined here
  https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#elf-marking

kovdan01 commented:
@jh7370 Thanks for your review, fixed the issues you've mentioned. Feel free to re-open resolved threads if something looks wrong for you.

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -5312,6 +5330,31 @@ static bool printAndroidNote(raw_ostream &OS, uint32_t NoteType,
+    return false;
+  }
+
+  uint64_t platform =
+      support::endian::read64<ELFT::TargetEndianness>(Desc.data() + 0);
+  uint64_t version =
```
These names don't conform with LLVM naming standards. Please address.

kovdan01 replied:
Fixed, thanks

jh7370 reviewed (commented):
Hi @kovdan01, two general process points:

1) Please avoid force pushes unless it's absolutely necessary to rebase your PR: it makes it impossible to view the changes since my previous PR, and potentially has other issues too (it's also covered by the existing LLVM review policy).
2) Please don't resolve coversations that I've initiated. I use this functionality as a checklist to ensure my comments have actually been addressed to my satisfaction, and if they're resolved, it makes it harder to track things. Please see my LLVM Discourse thread on the resolve conversation button for more details.

kovdan01 committed:
Address review comments

kovdan01 commented:
Thanks @jh7370 and @MaskRay for your feedback and comments. Addressed the issues in 8d7ee46a6bd46bacde19e4524ec7a33705ac3f7f - please consider reviewing the changes

jh7370 reviewed (approved):
Looks good from my point of view. Please give @MaskRay a chance to comment.

MaskRay commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -6208,6 +6263,13 @@ void ELFDumper<ELFT>::forEachRelocationDo(
     break;
+  case ELF::SHT_AARCH64_AUTH_RELR:
+    if (Obj.getHeader().e_machine != EM_AARCH64) {
+      this->reportUniqueWarning(
```
we should drop this for non-AArch64. They may define the value for other purposes. Just use the default code path to handle an unknown section header type.

kovdan01 replied:
I suppose a slightly different approach would be better if we have concerns regarding other targets defining the same section type for unrelated purposes. `ELFDumper<ELFT>::forEachRelocationDo` should only be called on sections which pass the check `isRelocationSec<LEFT>(Sec)`. So, we can change this check and make it take the target into account - as for now, it returns true for `SHT_AARCH64_AUTH_RESULT` unconditionally. After such change, we can just use an `assert(Obj.getHeader().e_machine == EM_AARCH64)` under `case ELF::SHT_AARCH64_AUTH_RESULT` in `ELFDumper<ELFT>::forEachRelocationDo` to ensure the contract is satisfied.

@MaskRay Please let me know your thoughts on this.

MaskRay replied:
I haven't checked, but if we can guarantee `assert(Obj.getHeader().e_machine == EM_AARCH64)`. It's better than `if (EM_AARCH64 && ELF::SHT_AARCH64_AUTH_RELR) do`

kovdan01 replied:
It turns out that we can't assume that only relocation sections are passed to `forEachRelocationDo` - there are code paths when other section types also get there, for example, `printSectionHeaders -> printRelocationsHelper -> forEachRelocationDo`. So, we must break here on non-AArch64 emachine (the switch does not contain a default branch, so break looks OK for me).

I also added checks against emachine in other places where appropriate and moved code/tests to target-specific parts where needed. @MaskRay Please see f9103d548b61a60f41e7529ea357b6ee2b8a0f66 and let me know if it is OK.

MaskRay replied:
Testing `EHeader.e_machine == EM_AARCH64 && SType == ELF::SHT_AARCH64_AUTH_RELR` is reasonable.

kovdan01 committed:
Use `#` instead of `//` for lit and FileCheck directives

kovdan01 commented on llvm/test/tools/llvm-readobj/ELF/AArch64/aarch64-feature-pauth.s:
```diff
@@ -0,0 +1,98 @@
+# RUN: llvm-readobj --notes tag-short.o | FileCheck --check-prefix LLVM-SHORT %s
+# RUN: llvm-readobj --notes tag-long.o  | FileCheck --check-prefix LLVM-LONG %s
+
+// LLVM-SHORT:      Notes [
```
Fixed in b0c9a289d25fdf721385f4df3568533c3fb79fb4, thanks

kovdan01 committed:
Process SHT_AARCH64_AUTH_RELR as target-specific section type
